File: leetcode/0503-next-greater-element-ii/solution.cpp
#include <vector>
#include <stack>
using namespace std;

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        int n = nums.size();
        vector<int> answer(n, -1);
        stack<int> st;  // indices, values in decreasing order

        for (int i = 0; i < 2 * n; i++) {
            int idx = i % n;
            while (!st.empty() && nums[idx] > nums[st.top()]) {
                answer[st.top()] = nums[idx];
                st.pop();
            }
            if (i < n) {
                st.push(idx);
            }
        }

        return answer;
    }
};

File: leetcode/0503-next-greater-element-ii/solution.py
from typing import List


class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        answer = [-1] * n
        stack = []  # indices, values in decreasing order

        for i in range(2 * n):
            idx = i % n
            while stack and nums[idx] > nums[stack[-1]]:
                answer[stack.pop()] = nums[idx]
            if i < n:
                stack.append(idx)

        return answer
